### PR TITLE
List repositories of the logged in User

### PR DESCRIPTION
**Pull request by sreecodeslayer**
* Calls GitLab projects v4 api on each visit to the page
* Renders the projects with pagination

### CODE REVIEW

File: app/helpers/repo_helper.rb [hunk 1]
```diff
@@ -0,0 +1,25 @@
+# frozen_string_literal: true
+
+module RepoHelper
```
Comment (thegeorgeous) [inline app/helpers/repo_helper.rb]: For this type of code, the ideal way is to use Services. Helpers are meant for the View layer.

File: app/controllers/repo_controller.rb [hunk 2]
```diff
@@ -0,0 +1,10 @@
+# frozen_string_literal: true
+
+class RepoController < ApplicationController
+  include RepoHelper
+  before_action :authenticate
+
+  def index
+    @projects, @pagination = get_projects_of_user(@current_user, params[:page])
```
Comment (thegeorgeous) [inline app/controllers/repo_controller.rb]: Using classes over methods is preferred in this case.
https://blog.engineyard.com/keeping-your-rails-controllers-dry-with-services

File: app/helpers/repo_helper.rb [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+  def get_projects_of_user(user, page)
+    all_user_projects_url = "https://gitlab.com/api/v4/users/#{user.gid}"\
+      "/projects?page=#{page}&per_page=9"
+    headers = {
+      'Accept' => 'application/json',
+      'Authorization' => "Bearer #{user.access_token}"
+    }
+    rsp = RestClient.get(all_user_projects_url, headers)
+    pagination = {
```
Comment (thegeorgeous) [inline app/helpers/repo_helper.rb]: Ideally, keep pagination and fetching projects into two different parts. Fetching projects is part of business logic. Pagination is view logic. That way we can adhere to SRP

File: app/views/layouts/_pagination.html.erb [hunk 4]
```diff
@@ -0,0 +1,4 @@
+<nav class="pagination" role="navigation" aria-label="pagination">
+  <a href="<%= repositories_path(:page => @pagination[:prev]) %>" class = "pagination-previous">Previous</a>
+  <a href="<%= repositories_path(:page => @pagination[:next]) %>" class = "pagination-next">Next</a>
+</nav>
```
Comment (thegeorgeous) [inline app/views/layouts/_pagination.html.erb]: Enable the 'add newline at end of file'  option in your editor. That will help avoid this sort of problems

File: Gemfile [hunk 5]
```diff
@@ -64,3 +64,5 @@ end
 gem 'tzinfo-data', platforms: %i[mingw mswin x64_mingw jruby]
 
 gem 'omniauth-gitlab', '~> 2.0'
+
+gem 'rest-client', '~> 2.1'
```
Comment (thegeorgeous) [inline Gemfile]: Good choice! 👍 

We might have to Faraday later though

File: app/helpers/repo_helper.rb [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+  require 'json'
+
+  def get_projects_of_user(user, page)
+    all_user_projects_url = "https://gitlab.com/api/v4/users/#{user.gid}"\
+      "/projects?page=#{page}&per_page=9"
+    headers = {
+      'Accept' => 'application/json',
+      'Authorization' => "Bearer #{user.access_token}"
+    }
+    rsp = RestClient.get(all_user_projects_url, headers)
```
Comment (thegeorgeous) [inline app/helpers/repo_helper.rb]: This needs a better name

File: app/helpers/repo_helper.rb [hunk 7]
```diff
@@ -0,0 +1,25 @@
+# frozen_string_literal: true
+
+module RepoHelper
+  require 'rest-client'
+  require 'json'
+
+  def get_projects_of_user(user, page)
+    all_user_projects_url = "https://gitlab.com/api/v4/users/#{user.gid}"\
```
Comment (thegeorgeous) [inline app/helpers/repo_helper.rb]: When you move this to a service, you could break this down into parts. Like, `base_url`, `version`, `endpoint`.
This will help us adapt to self-hosted Gitlab, different versions, different endpoints etc.

**Comment by thegeorgeous** [comment]
@sreecodeslayer, could you also some tests for this PR?

File: Gemfile (hunk 5 above)
Comment (sreecodeslayer) [reply Gemfile]: I saw them both, and thought there isn't much happening as of now. People vouched for Faraday where there is a need of more fine grade control.

File: app/helpers/repo_helper.rb (hunk 1 above)
Comment (sreecodeslayer) [reply app/helpers/repo_helper.rb]: Ah, that's what I was looking for. Will change 👍

File: app/helpers/repo_helper.rb (hunk 6 above)
Comment (sreecodeslayer) [reply app/helpers/repo_helper.rb]: rsp?

File: app/services/gitlab_repo_service.rb [hunk 8]
```diff
@@ -0,0 +1,35 @@
+# frozen_string_literal: true
+
+require 'rest-client'
+require 'json'
+require 'uri'
```
Comment (thegeorgeous) [inline app/services/gitlab_repo_service.rb]: Are these necessary?

File: app/services/gitlab_repo_service.rb [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+
+class GitlabRepoService
+  def initialize(user)
+    @base_url = 'https://gitlab.com/api/'
+    @api_version = 'v4/'
+    @user_projects_endpoint = "users/#{user.gid}/projects"
+    @per_page = 9
```
Comment (thegeorgeous) [inline app/services/gitlab_repo_service.rb]: Use initializer for setting state. Anything that is not related to the state of the class should be a constant

Comment (sreecodeslayer) [reply app/services/gitlab_repo_service.rb]: aha, been a while since I wrote classes! 😄
changed!

File: app/services/gitlab_repo_service.rb (hunk 8 above)
Comment (sreecodeslayer) [reply app/services/gitlab_repo_service.rb]: umm, no 🤔 
removed them
I was following the basic usage example from [rest-client readme](https://github.com/rest-client/rest-client#usage-raw-url). 

Is rails doing something here?

File: app/helpers/repo_helper.rb (hunk 6 above)
Comment (thegeorgeous) [reply app/helpers/repo_helper.rb]: yes

File: app/services/gitlab_repo_service.rb (hunk 8 above)
Comment (thegeorgeous) [reply app/services/gitlab_repo_service.rb]: Unless you add `require: false` next to a gem, it is usually available globally without needing a `require` line

Comment (sreecodeslayer) [reply app/services/gitlab_repo_service.rb]: I see. Got it

File: app/helpers/repo_helper.rb (hunk 6 above)
Comment (sreecodeslayer) [reply app/helpers/repo_helper.rb]: Already corrected :slightly_smiling:
